### PR TITLE
Update codegen to use boxed kernel

### PR DESCRIPTION
**Pull request by soulitzer**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #63878
* __->__ #63459
* #63458
 - Replaces the usual registration basically when "requires_derivative" is True (as in we still need a grad_fn), but `fn.info` is `None` (TODO maybe make sure differentiable inputs > 0 also to match requires_derivative).
 - Adds some (temporary?) fixes to some sparse functions See: https://github.com/pytorch/pytorch/issues/63549
 - To remove the codegen that generates NotImplemented node (though that should only be one line),  because there are some ops listed under `RESET_GRAD_ACCUMULATOR` that have a extra function call. We would need to make this list of ops available to c++, but this would either mean we'd have to codegen a list of strings, or move the RESET_GRAD_ACCUMULATOR to cpp land. We could do this in a future PR if necessary.

Differential Revision: [D30518571](https://our.internmc.facebook.com/intern/diff/D30518571)

### CODE REVIEW

File: tools/autograd/gen_variable_type.py [hunk 1]
```diff
@@ -404,13 +408,34 @@ def gen_variable_type_func(
...
+            # proceeds as usual but an error is properly produced on backward.
+            #
+            # There are two cases where still let codegen handle it:
+            # 1) ops that need to reset grad accumulator (we let codegen handle this case
+            #     because) the list is (currently) only accessible in Python and there are
+            #     only 2 op.
+            # 2) User explicitly specifies DONT_REQUIRE_DERIVATIVE. This basically makes
+            #    autograd a fallthrough with NDEBUG checks. This can be useful for when all
+            #    outputs are integral.
+            # 3) When there are no differentiable outputs. This is similar to (2).
```
Comment (ezyang) [inline tools/autograd/gen_variable_type.py:427]: Yeah it would be nice to nail these too, somehow. Not sure what the best way is.

File: aten/src/ATen/native/sparse/SoftMax.cpp [hunk 2]
```diff
@@ -407,6 +407,9 @@ void cpu_sparse_coo_softmax_backward(const Tensor& grad_input, const Tensor& gra
   indices.resize_as_(out_indices);
   indices.copy_(out_indices);
 
+  // TOOO: We need this when we call into from RedispatchBoxed (why?)
+  at::sparse::alias_into_sparse(grad_input, indices, values);
```
Comment (ezyang) [inline aten/src/ATen/native/sparse/SoftMax.cpp:411]: Yeah, I don't really understand why this is necessary either. Would be good to understand better.

**Review by ezyang — APPROVED** [review]

File: aten/src/ATen/native/sparse/SoftMax.cpp (hunk 2 above)
Comment (soulitzer) [reply aten/src/ATen/native/sparse/SoftMax.cpp:411]: The issue is that somewhere in the dispatcher a new TensorImpl is being created, making it so that resize_as_ on the returned _values does not affect the original sparse tensor.

So far I've found that TensorImpl returned by `_values` is TensorImpl returned by `op.call(self)` is different from that of `at::native::_values_sparse` (but only in certain situations), so the problem could be somewhere in the dispatcher.
 - Register in `native_functions.yaml` (no new TensorImpl)
 - Register in cpp extension (new TensorImpl) breaks
 - Register in `native_functions.yaml`, but call w/ redispatchBoxed, i.e., when we have this boxed fallback

I've tried a bit to bisect exactly where it occurs, but its surprisingly difficult to print trivial things in the dispatcher code: see https://github.com/pytorch/pytorch/pull/63652/files.

Comment (soulitzer) [reply aten/src/ATen/native/sparse/SoftMax.cpp:411]: cc @albanD

Comment (ezyang) [reply aten/src/ATen/native/sparse/SoftMax.cpp:411]: Can you verify the that new code (in the boxed fallback) is equivalent to the code that would have been generated for sparse tensors? That seems like the most likely cause of the divergence. A good start is to paste the before code for this function.

Comment (soulitzer) [reply aten/src/ATen/native/sparse/SoftMax.cpp:411]: Good call, there was indeed a difference here. (AutoDispatchBelowADInplaceOrView <-> AutoDispatchBelowAutograd)
The issue is that we are dispatching into the InplaceOrView kernel for _values/_indices causing it to create a new TensorImpl in as_view. So a simple fix until we have our new InplaceOrView boxed kernel is have it AutoDispatchBelowADInplaceOrView always instead of AutoDispatchBelowAutograd.

Comment (ezyang) [reply aten/src/ATen/native/sparse/SoftMax.cpp:411]: Great!

Comment (ezyang) [reply aten/src/ATen/native/sparse/SoftMax.cpp:411]: So you're going to undo these changes now right? :)

**Comment by soulitzer** [comment]
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30518571).
